### PR TITLE
Apply again email content

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

Now we are building apply again, candidates need to be told about it when all of their options are rejected/withdrawn/declined.

This PR creates new emails for the above situations informing them about apply again.

## Changes proposed in this pull request

- Update the content of the all rejected email
- Create 2 new DBD emails, 1 when they have a rejected course choice and 1 without, informing the candidate of apply2
- Adds an email for when a candidate withdraws an application choice and has none which are successful
- Adds an email for when a candidate declines an application choice and has none which are successful

All rejected 

Before

image

After

image

DBD email - all withdrawn or declined

image

DBD email - all rejected/withdrawn/declined with at least 1 rejection

image

Withdraw email

image

Decline email 

image

## Guidance to review

Use the mailer previews to check the emails have the correct content

content document - https://docs.google.com/document/d/14S_40eWV8MYZnjGoW1Pop4LfdXuMXkrHRlilLpoBDdw/edit

## Link to Trello card

https://trello.com/c/uSMiuUHp/1388-update-email-content-to-allow-candidates-to-apply-again

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/feature_flag.rb [hunk 1]
```diff
@@ -8,6 +8,7 @@ class FeatureFlag
   ].freeze
 
   TEMPORARY_FEATURE_FLAGS = %w[
+    apply_again_email_content
```
Comment (tvararu) [inline app/services/feature_flag.rb:11]:
```suggestion
```

Let's reuse the `apply_again` flag 👍

File: app/views/candidate_mailer/application_rejected_all_rejected.erb [hunk 2]
```diff
@@ -1,3 +1,37 @@
+<% if FeatureFlag.active?('apply_again_email_content') %>
```
Comment (tvararu) [inline app/views/candidate_mailer/application_rejected_all_rejected.erb:1]:
```suggestion
<% if FeatureFlag.active?('apply_again') %>
```

File: app/views/candidate_mailer/declined_by_default.text.erb [hunk 3]
```diff
@@ -1,3 +1,55 @@
+<% if FeatureFlag.active?('apply_again_email_content') && @last_course_choice_and_none_successful %>
```
Comment (tvararu) [inline app/views/candidate_mailer/declined_by_default.text.erb:1]:
```suggestion
<% if FeatureFlag.active?('apply_again') && @last_course_choice_and_none_successful %>
```

File: app/views/candidate_mailer/application_rejected_all_rejected.erb [hunk 4]
```diff
@@ -1,3 +1,37 @@
...
+
+<%= @course.provider.name %> has decided not to progress your teacher training application for <%= @course.name_and_code %> on this occasion.
+
+# You may still be able to get a place on a course
+
+Your <%= @application_form.application_choices.count > 1 ? 'provider has' : 'providers have' %> left you some feedback. There may be things you can do to get a place on a course this year, next year or in the future.
+
+Review your feedback:
+
+https://www.apply-for-teacher-training.education.gov.uk/candidate/account
```
Comment (fofr) [inline app/views/candidate_mailer/application_rejected_all_rejected.erb:15]: Are we no longer showing the feedback in our rejection email? cc @EmmaFrith @paulrobertlloyd

File: app/views/candidate_mailer/declined_by_default.text.erb [hunk 5]
```diff
@@ -1,3 +1,55 @@
...
+
+# <%= 'Application'.pluralize(@declined_course_names.size) %> withdrawn automatically
+
+<% if FeatureFlag.active?('covid_19') %>
+You did not respond in time so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>.
+<% else %>
+You did not respond within <%= @declined_courses.first.decline_by_default_days %> working days so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>
+<% end %>
+
+<% if @rejected_course_count.zero? %>
```
Comment (fofr) [inline app/views/candidate_mailer/declined_by_default.text.erb:13]: I wonder if we can make this logic clearer that this block is for anyone who hasn't received a rejection. Took me a few readings to parse.

File: app/views/candidate_mailer/application_rejected_all_rejected.erb (hunk 4 above)
Comment (EmmaFrith) [reply app/views/candidate_mailer/application_rejected_all_rejected.erb:15]: This is looking awesome thanks @davidgisbey!

@fofr @paulrobertlloyd - I took the view to not include the reasons for rejection in this final email because a) the main thing we want them to do is to log back in and apply again b) we want them to review all their feedback as a whole to see if they're suitable for applying again c) the email is already long and we don't want them to miss the 'apply again' call to action. Does this sound sensible?

File: app/services/feature_flag.rb [hunk 6]
```diff
@@ -9,6 +9,7 @@ class FeatureFlag
   ].freeze
 
   TEMPORARY_FEATURE_FLAGS = %w[
+    apply_again
```
Comment (tvararu) [inline app/services/feature_flag.rb:12]:
```suggestion
```

It's already present at the end of the list.

File: app/views/candidate_mailer/declined_by_default_with_rejections.html.erb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+# <%= 'Application'.pluralize(@declined_course_names.size) %> withdrawn automatically
+
+<% if FeatureFlag.active?('covid_19') %>
+You did not respond in time so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>.
+<% else %>
+You did not respond within <%= @declined_courses.first.decline_by_default_days %> working days so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>
+<% end %>
+
+#Are you still interested in teaching?
```
Comment (tvararu) [inline app/views/candidate_mailer/declined_by_default_with_rejections.html.erb:11]:
```suggestion
# Are you still interested in teaching?
```

Wouldn't nit for this but not sure if it works without the space ¯\\_(ツ)\_/¯

File: app/views/candidate_mailer/declined_by_default_without_rejections.html.erb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+# <%= 'Application'.pluralize(@declined_course_names.size) %> withdrawn automatically
+
+<% if FeatureFlag.active?('covid_19') %>
+You did not respond in time so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>.
+<% else %>
+You did not respond within <%= @declined_courses.first.decline_by_default_days %> working days so we declined your <%= 'offer'.pluralize(@declined_course_names.size) %> for <%= @declined_course_names.to_sentence %>
+<% end %>
+
+#Are you still interested in teaching?
```
Comment (tvararu) [inline app/views/candidate_mailer/declined_by_default_without_rejections.html.erb:11]:
```suggestion
# Are you still interested in teaching?
```

File: app/views/candidate_mailer/declined_by_default_without_rejections.html.erb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+Sign in to apply again:
+https://www.apply-for-teacher-training.education.gov.uk/candidate/account
+
+# Get help
+
+For advice on becoming a teacher, call Get Into Teaching for free on 0800 389 2500 or chat to an adviser online:
+https://getintoteaching.education.gov.uk/contact
+
+For any technical issues using Apply for teacher training, contact us at [becomingateacher@digital.education.gov.uk](mailto:becomingateacher@digital.education.gov.uk)
```
Comment (tvararu) [inline app/views/candidate_mailer/declined_by_default_without_rejections.html.erb:25]: Some of this content can be turned into partials and reused across multiple emails. Could we try that?

Comment (davidgisbey) [reply app/views/candidate_mailer/declined_by_default_without_rejections.html.erb:25]: Sure i'm going to do it in the next PR and merge them together

File: app/views/candidate_mailer/application_rejected_all_rejected.erb [hunk 10]
```diff
@@ -1,3 +1,37 @@
+<% if FeatureFlag.active?('apply_again') %>
```
Comment (tijmenb) [inline app/views/candidate_mailer/application_rejected_all_rejected.erb:1]: We should use separate templates for these too, it's quite tricky to see where the email starts/stops.

File: app/views/candidate_mailer/_interested_without_rejections.erb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Are you still interested in teaching?
+
+You didn’t pursue your teacher training application, but it’s not too late to apply again for a course if you’ve changed your mind.
+
+We’ve saved your last application, so all you have to do is choose a course, review your application and re-submit.
+
+Sign in to apply again:
+
+https://www.apply-for-teacher-training.education.gov.uk/candidate/account
```
Comment (tijmenb) [inline app/views/candidate_mailer/_interested_without_rejections.erb:9]: These should should the URL helpers, and include a magic link.

Comment (davidgisbey) [reply app/views/candidate_mailer/_interested_without_rejections.erb]: 👍

File: app/mailers/candidate_mailer.rb [hunk 12]
```diff
@@ -128,6 +128,20 @@ def declined_by_default(application_form)
...
+    email_for_candidate(application_form, subject: I18n.t!('candidate_mailer.decline_by_default_last_course_choice.subject', count: @declined_courses.size))
+  end
+
+  def declined_by_default_without_rejections(application_form)
+    @declined_courses = application_form.application_choices.select(&:declined_by_default?)
+    @declined_course_names = @declined_courses.map { |application_choice| "#{application_choice.course_option.course.name_and_code} at #{application_choice.course_option.course.provider.name}" }
+
+    email_for_candidate(application_form, subject: I18n.t!('candidate_mailer.decline_by_default_last_course_choice.subject', count: @declined_courses.size))
+  end
+
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/mailers/candidate_mailer.rb:145]: Okay. I'll give it a go. I'm a bit concerned that it's going to be really hard to parse. I'll get it working and try and keep it as simple as possible.

File: spec/mailers/candidate_mailer_offers_and_rejections_spec.rb [hunk 13]
```diff
@@ -121,13 +121,22 @@ def setup_application
...
 
     describe '.application_rejected_all_rejected' do
-      it_behaves_like(
-        'a mail with subject and content', :application_rejected_all_rejected,
-        I18n.t!('candidate_mailer.application_rejected.all_rejected.subject', provider_name: 'Falconholt Technical College'),
-        'heading' => 'Dear Tyrell',
-        'course name and code' => 'Forensic Science (E0FO)',
-        'providers rejection reason' => 'The application had little detail.'
-      )
+      if FeatureFlag.active?('apply_again')
```
Comment (tijmenb) [inline spec/mailers/candidate_mailer_offers_and_rejections_spec.rb:124]: Is this correct? What will have activated/disabled the feature flag here?

Comment (davidgisbey) [reply spec/mailers/candidate_mailer_offers_and_rejections_spec.rb]: yea that absolutely should not be like that. It should test with and without the flag on. I've fixed it.

File: app/views/candidate_mailer/application_rejected_all_rejected_apply_again.erb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+<%= @candidate_magic_link %>
+
+# Get help
+
+We’re here to help you get a place on a course. Call Get Into Teaching for free on 0800 389 2500 or chat to an adviser online:
+
+https://getintoteaching.education.gov.uk/contact
+
+For any technical issues using Apply for teacher training, contact us at becomingateacher@digital.education.gov.uk
```
Comment (tvararu) [inline app/views/candidate_mailer/application_rejected_all_rejected_apply_again.erb:29]:
```suggestion
For any technical issues using Apply for teacher training, contact us at [becomingateacher@digital.education.gov.uk](mailto:becomingateacher@digital.education.gov.uk)
```

File: app/views/candidate_mailer/decline_last_application_choice.text.erb [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+You’ve declined your offer to study <%= @declined_course_name %>.
+
+<% if @rejected_course_choices_count.zero? %>
+  <%= render 'interested_without_rejections' %>
+  <%= render 'get_help' %>
+<% else %>
+  <%= render 'interested_with_rejections' %>
+  <%= render 'get_help' %>
+<% end %>
```
Comment (tvararu) [inline app/views/candidate_mailer/decline_last_application_choice.text.erb:13]:
```suggestion
<% if @rejected_course_choices_count.zero? %>
  <%= render 'interested_without_rejections' %>
<% else %>
  <%= render 'interested_with_rejections' %>
<% end %>
<%= render 'get_help' %>
```

Comment (tvararu) [reply app/views/candidate_mailer/decline_last_application_choice.text.erb:13]: I think these templates look a lot better now with partials 👍

File: app/views/candidate_mailer/withdraw_last_application_choice.text.erb [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+You’ve withdrawn your <%= 'application'.pluralize(@withdrawn_course_names.size) %> for <%= @withdrawn_course_names.to_sentence %>.
+
+<% if @rejected_course_choices_count.zero? %>
+  <%= render 'interested_without_rejections' %>
+  <%= render 'get_help' %>
+<% else %>
+  <%= render 'interested_with_rejections' %>
+  <%= render 'get_help' %>
+<% end %>
```
Comment (tvararu) [inline app/views/candidate_mailer/withdraw_last_application_choice.text.erb:13]:
```suggestion
<% if @rejected_course_choices_count.zero? %>
  <%= render 'interested_without_rejections' %>
<% else %>
  <%= render 'interested_with_rejections' %>
<% end %>
<%= render 'get_help' %>
```

File: app/views/candidate_mailer/application_rejected_all_rejected_apply_again.erb (hunk 14 above)
Comment (paulrobertlloyd) [reply app/views/candidate_mailer/application_rejected_all_rejected_apply_again.erb:29]: Or even `<becomingateacher@digital.education.gov.uk>`
